datafusion/common/src/file_options/mod.rs

```diff
@@ -0,0 +1,269 @@
+    }
+
+    pub fn into_inner(self) -> Vec<OptionTuple> {
+        self.options
+    }
+
+    /// Scans for option and if it exists removes it and attempts to parse as a boolean
+    /// Returns none if it does not exist.
+    pub fn get_bool_option(&mut self, find: &str) -> Result<Option<bool>> {
+        let maybe_option = self.scan_and_remove_option(find);
+        maybe_option
+            .map(|(_, v)| parse_boolean_string(find, v))
+            .transpose()
+    }
+
+    /// Scans for option and if it exists removes it and returns it
+    /// Returns none if it does not exist
+    pub fn get_str_option(&mut self, find: &str) -> Option<String> {
+        let maybe_option = self.scan_and_remove_option(find);
+        maybe_option.map(|(_, v)| v)
+    }
+
+    /// Infers the file_type given a target and arbitrary options.
+    /// If the options contain an explicit "format" option, that will be used.
+    /// Otherwise, attempt to infer file_type from the extension of target.
+    /// Finally, return an error if unable to determine the file_type
+    /// If found, format is removed from the options list.
+    pub fn try_infer_file_type(&mut self, target: &str) -> Result<FileType> {
+        let explicit_format = self.scan_and_remove_option("format");
+        let format = match explicit_format {
+            Some(s) => FileType::from_str(s.1.as_str()),
+            None => {
+                // try to infer file format from file extension
+                let extension: &str = &Path::new(target)
+                    .extension()
+                    .ok_or(DataFusionError::InvalidOption(
+                        "Format not explicitly set and unable to get file extension!"
+                            .to_string(),
+                    ))?
+                    .to_str()
+                    .ok_or(DataFusionError::InvalidOption(
+                        "Format not explicitly set and failed to parse file extension!"
+                            .to_string(),
+                    ))?
+                    .to_lowercase();
+
+                FileType::from_str(extension)
+            }
+        }?;
+
+        Ok(format)
+    }
+
+    /// Finds an option in StatementOptions if exists, removes and returns it
+    /// along with the vec of remaining options.
+    fn scan_and_remove_option(&mut self, find: &str) -> Option<OptionTuple> {
+        let idx = self
+            .options
+            .iter()
+            .position(|(k, _)| k.to_lowercase() == find.to_lowercase());
+        match idx {
+            Some(i) => Some(self.options.swap_remove(i)),
+            None => None,
+        }
+    }
+}
+
+/// This type contains all options needed to initialize a particular
+/// RecordBatchWriter type. Each element in the enum contains a thin wrapper
+/// around a "writer builder" type (e.g. arrow::csv::WriterBuilder)
+/// plus any DataFusion specific writing options (e.g. CSV compression)
+#[derive(Clone, Debug)]
+pub enum FileTypeWriterOptions {
```

Comment:
This enum was proposed by @alamb in #6569 and is the key new abstraction in this PR.